### PR TITLE
[QEff.finetuning] Hf config update

### PR DESCRIPTION
**Pull request by tchawada**
Added QAIC validation to config_manager.py and introduced a default value for prompt_func in the Alpaca dataset configuration. Updated same in documentation.

To run integrated_test for DDP use following command:
QAIC_VISIBLE_DEVICES=0,1 torchrun --nproc-per-node=2 -m pytest -q QEfficient/finetune/experimental/tests/test_integrated.py

### CODE REVIEW

File: QEfficient/finetune/experimental/core/config_manager.py [hunk 1]
```diff
@@ -24,12 +24,35 @@
 logger = Logger(__name__)
 
 
+def is_NSP_free():
```
Comment (quic-akuruvil) [inline QEfficient/finetune/experimental/core/config_manager.py]: Func looks fine, please put this func inside some utils file.

File: QEfficient/finetune/experimental/core/dataset.py [hunk 2]
```diff
@@ -117,8 +113,7 @@ def _initialize_dataset(self):
         """
         if self.json_file_path:
             # Load dataset from JSON file
-            self.dataset = load_dataset("json", data_files=self.json_file_path, split="train")
-
+            self.dataset = load_dataset("json", data_files=self.json_file_path, split=self.split)
             # Apply train/test split if needed
             if self.split in ["train", "test"]:
```
Comment (quic-swatia) [inline QEfficient/finetune/experimental/core/dataset.py]: Please remove this as discussed, also observe the diff  in number of samples based on this in both alpaca and gsm8k dataset.

File: QEfficient/finetune/experimental/tests/test_integrated.py [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+    TEST_SEED,
+    TEST_WARMUP_STEPS,
+    TEST_WEIGHT_DECAY,
+    TEST_LOSS_THRESHOLD,
+    AutoClassName,
+    DatasetType,
+    TaskType,
+)
+from QEfficient.finetune.experimental.core.utils.peft_utils import convert_peft_config_to_lora_config
+from QEfficient.finetune.experimental.core.utils.training_config_utils import prepare_training_config
```
Comment (quic-akuruvil) [inline QEfficient/finetune/experimental/tests/test_integrated.py]: Please resolve this, use `ruff check --fix` or `ruff format --fix`

**Review by quic-akuruvil — COMMENTED** [review]
Please add tests for multinode setup as well

File: QEfficient/finetune/experimental/tests/test_integrated.py [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+        train_result: Training result object
+        eval_result: Evaluation result dictionary
+    """
+    assert train_result is not None
+    assert hasattr(train_result, "training_loss")
+    assert "eval_loss" in eval_result
+    logger.warning(f"Training loss: {train_result.training_loss:.4f}")
+    logger.warning(f"Evaluation loss: {eval_result['eval_loss']:.4f}")
+    assert abs(train_result.training_loss - eval_result["eval_loss"]) < TRAIN_EVAL_EPOCH_LOSS_DIFF_THRESHOLD
+
```
[the remark this answers is not in the document]
Comment (quic-akuruvil) [reply QEfficient/finetune/experimental/tests/test_integrated.py]: https://github.com/quic/efficient-transformers/blob/a6f2dd410faaf8afff89276a5a6d4d359254015e/tests/finetune/test_finetune.py#L322

File: QEfficient/finetune/experimental/tests/test_config.yaml [hunk 5]
```diff
@@ -12,9 +12,8 @@ model:
   model_name: "HuggingFaceTB/SmolLM-135M"  # Pretrained model name
   use_peft: true
   peft_config:
-    lora_r: 8
-    lora_alpha: 16
-    lora_dropout: 0.1
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/tests/test_config.yaml]: If we are removing this then default dropout of 0.1 will be applied?

File: QEfficient/finetune/experimental/tests/test_integrated.py [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+        train_dataset: Train dataset instance
+        eval_dataset: Evaluation dataset instance
+        optimizer_cls_and_kwargs: Optimizer class and kwargs
+        training_args: Training arguments
+        callbacks: List of callback instances
+        pipeline: FineTuningPipeline instance
+    Returns:
+        Trainer instance
+    """
+    trainer = pipeline._create_trainer(
```
Comment (quic-meetkuma) [inline QEfficient/finetune/experimental/tests/test_integrated.py]: Ideally we should not call any internal methods (_create_trainer) in test cases. If such methods are required then make them public or modify the tests so that it will not use internal (private) methods.

File: QEfficient/utils/device_utils.py [hunk 7]
```diff
@@ -21,6 +23,31 @@ def is_networks_loaded(stdout):
     return False
 
 
+def is_nsp_free():
+    # FIXME: Give incorrect results when user doesn't have permission.
```
Comment (quic-meetkuma) [inline QEfficient/utils/device_utils.py]: This function should only be called by one rank/device, not by all the ranks. Because it contains os.system calls and it has for loop over all the devices.

File: QEfficient/finetune/experimental/tests/test_config.yaml (hunk 5 above)
Comment (tchawada) [reply QEfficient/finetune/experimental/tests/test_config.yaml]: Yes

File: QEfficient/cloud/finetune_experimental.py [hunk 8]
```diff
@@ -197,14 +198,23 @@ def _create_trainer(
 
         # Create trainer arguments instance
         args = args_cls(**training_config)
-        # Initialize trainer
+        dataset_config_dict = self.config_manager.get_dataset_config()
+        split_ratio = dataset_config_dict.get("split_ratio", 0.8)
+        num_samples = dataset_config_dict.get("dataset_num_samples", -1)
+        train_dataset = train_dataset.dataset
+        eval_dataset = eval_dataset.dataset
+        if num_samples > 0:
```
Comment (quic-akuruvil) [inline QEfficient/cloud/finetune_experimental.py]: Please raise a warning saying for eg: ( fewer samples affect the quality of finetuning) if num_samples is too less, compared to total dataset samples. Also check if num_samples is non-negative in tests.
